Add some utility functions for obtaining runtime-relative paths.

al45tair opened the pull request:
Adds the ability to locate files relative to the path of the Swift runtime library, libswiftCore.

rdar://103071801

al45tair committed:
Add some utility functions for obtaining runtime-relative paths.

Adds the ability to locate files relative to the path of the Swift runtime
library, libswiftCore.

rdar://103071801

al45tair commented:
@swift-ci Please smoke test

compnerd commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,196 @@
+  const DWORD dwBufSize = 4096;
+  LPSTR lpFilename = (LPSTR)std::malloc(bufsize);
+
+  DWORD dwRet = GetMappedFileNameA(GetCurrentProcess(),
```
We really need to use `GetMappedFileNameW`.  However, that is not going to be a realpath.  We should use `GetFinalPathNameByHandleW` taking the module handle to map it to the realpath.  Ascii paths may be invalid if the codepoint cannot be mapped into the ASCII plane.

al45tair replied:
I did wonder about that, and whether or not that's genuinely going to be a problem.

That then begs a couple more questions:

1. Do we want this to return UTF-8 on Windows. It could do that, and that would play nice with Swift if anyone wanted to call these functions from Swift code. _However_ it also wouldn't work with the ANSI Windows APIs (including the C library, right?) unless someone has forced the code page to UTF-8.
2. Alternatively, we could have it return a wide string on Windows — but that presumably won't play so nicely with Swift code, and the API will then not match that on other platforms.

compnerd replied:
I think that the right thing to do is what we do elsewhere: system uses UTF-16 unconditionally.  Going from system to Swift, we convert and convert again from Swift to system.

al45tair replied:
My only concern here is that calling `open()` from Swift on Windows with a path derived from one of these functions is then not guaranteed to work without setting the code page to UTF-8. I suppose that's a general hazard on Windows, but it mightn't be obvious to the uninitiated.

al45tair replied:
Hmmm. I don't think `GetFinalPathNameByHandleW()` takes an `HMODULE`. In fact, I'm fairly sure it won't (it's been a while so I might be misremembering, but an `HMODULE` is really a pointer to the `IMAGE_DOS_HEADER`, IIRC, rather than a file handle which is what `GetFinalPathNameByHandleW()` accepts).

We _could_ use `GetModuleFileNameExW()` rather than `GetMappedFileNameW()`, but that means we'd need the `HMODULE` for `libswiftCore.dll` — admittedly that's rather easy to get (via `__ImageBase`), but I'm not sure we gain that much in practice.

al45tair replied:
Or are you suggesting that we should use `CreateFile()` to open the path we find, and pass that through `GetFinalPathNameByHandleW()`? I can see _an_ argument for doing something like that, and similarly calling `realpath()` on UNIX, in that it would mean you could symlink just `libswiftCore.dylib` and then resolve auxiliary items relative to the _target_ of the symlink. But equally, that would mean that you _couldn't_ symlink it and then resolve items relative to the location of the symbolic link, which is potentially equally useful (e.g. you could then replace just the auxiliary data).

al45tair replied:
(On the `GetModuleFileNameExW()` versus `GetMappedFileNameW()` front, I note that the docs for the former claim it may fail if the module list is not yet initialised or if it changes while the function is being called. There is no such note for the latter, so it might in fact be more robust.)

al45tair replied:
I now understand why you thought `GetFinalPathNameByHandleW()` would be needed. It's not _quite_ the right function though, and to use it we'd have to use `NtCreateFile()` which I'm reluctant to do because that's an internal API. So I've solved it in a similar way to the code MS gives for mapping these filenames on previous Windows versions.

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,196 @@
+#define EXE_EXT ""
+#endif
+
+}
```
I have no particular objection to using `<filesystem>` if you think that's the best way.

compnerd replied:
I think that we do have that at our disposal (it is part of the C++ standard, and greatly simplifies the path handling).

al45tair replied:
So, it turns out that we _can't_ use `<filesystem>` after all, because of the deployment target for the runtime on Darwin systems (it's implemented, but it isn't available until macOS 10.15 or above).

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,196 @@
+                      "libswiftCore should be installed in a `lib` "
+                      "directory: %s\n",
+                      runtimePath);
+  }
```
(This is exactly why I asked you to review it :-))

So, on Windows, if we want to get at another file from the runtime, where would it be relative to the runtime? The point of having these functions is to allow us to do that in a system-independent manner, so if Windows needs to do something slightly different, that's just fine.

compnerd replied:
I think that is the problem, the default installation does use a unix-ish layout, but that isn't guaranteed.  It should support the flat tree (application co-installed redistribution) and the system installed layout.

mikeash replied:
Seems like we should tolerate anything, and fail gracefully if we can't understand the layout we're in.

al45tair replied:
I've made it more flexible. If it finds `libswiftCore` in a `lib` directory, it will remove the `lib` part as well, otherwise it just returns the directory it found `libswiftCore` in.

mikeash commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,196 @@
+                      "Unable to obtain Swift runtime path\n");
+  }
+
+  runtimePath = ::strdup(dli.dli_fname);
```
It really seems like this string should already be immortal and we shouldn't need to copy it. But I can't find any guarantees.

al45tair replied:
I didn't want to make that assumption. It probably is immortal but for the sake of a one-off copy I'm not too worried. I have tweaked the code in this area a bit, however.

mikeash commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,196 @@
+_swift_initRuntimePath(void *) {
+  Dl_info dli;
+
+  int ret = ::dladdr((void *)_swift_initRuntimePath, &dli);
```
This should use the SPI `dyld_image_path_containing_address` where available, as it's slightly more efficient. There's precedent at https://github.com/apple/swift/blob/0213f2dd884ec4286d3aee03528cbfc6a961d54c/stdlib/public/runtime/ObjCRuntimeGetImageNameFromClass.mm#L82

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,196 @@
+_swift_initRootPath(void *)
+{
+  // SWIFT_ROOT overrides the path returned by this function
+  char *swiftRoot = getenv("SWIFT_ROOT");
```
I don't mind fixing it here… I've had a go at that — it's a bit grotty because I wanted the help output to say "string", which necessitates some unfortunate `typedef`s/macros. Not sure what you think about that, but take a look.

mikeash replied:
That is kind of ugly, but that's to be expected with crazy macros. Would using `std::string` as the backing type help at all? Otherwise, seems OK as is.

al45tair replied:
I suppose it might, if we did a `using string = std::string` or similar, perhaps.

al45tair replied:
No. Using `std::string` results in a need for global destructors. Might be able to get rid of the `string` macro in `EnvironmentVariabes.h` though — we're inside a namespace so we should be good to declare `string` locally in there.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test Windows platform

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

grynspan commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,276 @@
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+void
+_swift_initRuntimePath(void *) {
```
Could you use `SymbolInfo` here? It would eliminate the platform-specific shenanigans.

al45tair replied:
In principle, though we'd still want to call `dyld_image_path_containing_address()` on Darwin, and `SymbolInfo` uses `DbgHelp` on Windows, right, whereas this doesn't.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
Fix the paths up a little, and address a linking issue on Windows.

The path to the runtime should be something like

  /usr/lib/swift/libswiftCore.dylib

So we want to strip lib/swift/ off to get the Swift root.

We also want auxiliary executables in

  /usr/libexec/swift

rather than just

  /usr/libexec

(While still supporting flat layout for Windows.)

rdar://103071801

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test Windows platform

al45tair commented:
@swift-ci Please smoke test Windows platform

al45tair commented:
@swift-ci Please smoke test Windows platform

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
Tweak the linker path for Windows slightly.

Apparently the `swiftCore.lib` file is in `%target-sdk-name/%target-arch`
within the build directory.

rdar://103071801

al45tair committed:
Fix the paths test on Windows.

Windows doesn't have a "lib" prefix on the Swift DLLs.

rdar://103071801

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
Add code to map NT-style paths to Win32 paths.

GetMappedFilenameW() returns an NT-style path, not a Win32 path, but
we need a Win32 path.

rdar://103071801

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

compnerd commented on include/swift/Runtime/Paths.h:
```diff
@@ -0,0 +1,66 @@
+///         `SWIFT_ROOT` environment variable if set.
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRootPath();
```
I think that we should consider renaming this to `swift_getSwiftInstallRootPath`.

al45tair replied:
I think that might be misleading on Windows, because you might expect it to return the path to an installed copy of Swift, which if I've understood what you've told me about shipping Swift programs on Windows correctly, it definitely won't do (instead, it'll return the application's path, right?)

compnerd commented on include/swift/Runtime/Paths.h:
```diff
@@ -0,0 +1,66 @@
+/// \return A string containing the full path to the executable.
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getAuxiliaryExecutablePath(const char *name);
```
How about changing this to `swift_getPathForAuxiliaryExecutable`?

al45tair replied:
If it was on its own, I agree that would make sense, but it doesn't match the pattern for the other names, and I personally don't much like `swift_getPathForRootOfSwiftInstall` or `swift_getPathForRuntime`.

compnerd commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+  size_t extLen = 0;
+
+#ifdef _WIN32
+  if (nameLen > 4 && strcmp(name + nameLen - 4, ".exe") != 0)
```
Does it make more sense to use `PathCchFindExtension`?  I worry about silly things like XCTest which uses `.xctest` on Windows (yes, really).

al45tair replied:
I don't think it'll be a problem in this instance; the point of this was to let us write the name without the extension and have the Windows code add `.exe` automatically. I could have done it unconditionally, but I figured there was always a chance someone would pass something with `.exe` to it.

I don't think we want this except for executables (i.e. I don't think we should be using it for `.dll`s, because then it'll get much more complicated as they're `.so`s on Linux, `.dylib`s on macOS and potentially have other extensions on other platforms too).

al45tair commented on include/swift/Runtime/EnvironmentVariables.h:
```diff
@@ -24,6 +24,10 @@ void initialize(void *);
 
+// Define a typedef "string" in swift::runtime::environment to make string
+// environment variables work
+using string = const char *;
```
Yep. It is horrible. I blame whoever made all those fancy macros for the environment code.

mikeash replied:
Me too.

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+  while (ptr > runtimePath && !_swift_isPathSep(*--ptr));
+
+  // Remove lib/swift/ if present
+  if (ptr - runtimePath >= 10
```
I'm not hugely worried about that; `swift_getRuntimePath()` doesn't return anything that might come directly from user input, and that's the only real source of paths with runs of separators. It's a good thing to point out, mind.

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+    if (_swift_isPathSep(swiftRoot[len - 1])) {
+      rootPath = swiftRoot;
+    } else {
+      char *thePath = (char *)malloc(len + 2);
```
Yes. Using `new` in the runtime causes problems, because people can override `operator new` and since some of our code gets run really early, their `operator new` implementation might not be completely set-up at this point. We've had crashes caused by exactly that problem.

I don't think there's much benefit to using `strlcpy()` here, and I'm not sure how you'd use `strncat()` with `strdup()` (since the latter is going to allocate space just for the original string). `sprintf()` or `snprintf()` are an option too, but we already know the length of the string at this point, because we literally just measured it.

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+/// We can't use GetFinalPathNameByHandle() because there's no way to obtain
+/// a handle (at least, not without using the internal NtCreateFile() API, which
+/// we aren't supposed to be using).  Additionally, that function would resolve
+/// symlinks, which we don't want to do here.
```
I'm not sure why you would do that. `__ImageBase` is already the module handle; that is, an `HMODULE` just points at the image base.

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+  if (!dwLen)
+    return NULL;
+
+  LPWSTR lpDriveStrings = (LPWSTR)std::malloc(dwLen * sizeof(WCHAR));
```
Since we don't want to use `new`, using `std::unique_ptr` becomes a bit messier.

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+        size_t driveLen = wcslen(pszDrive);
+        size_t pathLen = fnLen - ntLen;
+        size_t newLen = driveLen + pathLen + 1;
+        LPWSTR pszWin32Name = (LPWSTR)std::malloc(newLen * sizeof(WCHAR));
```
See previous comment.

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+{
+  const char *rootPath = swift_getRootPath();
+
+  // Form <rootPath>/libexec/swift/
```
So, with the layout you have on Windows, I'd expect `swift_getRootPath()` to point at the `bin` directory in the install, which would then mean that `swift_getAuxiliaryExecutablePath()` will look for `bin\libexec\swift`, which doesn't exist, so will fall back to just looking in the `bin` directory.

Or, obviously, if someone ships an application, `swift_getRootPath()` will point to the application's directory, and the same thing will happen.

Were you thinking that `swift_getRootPath()` should strip the `bin`? In which case `swift_getAuxiliaryExecutablePath()` would need to look in `<swift root>\bin` on Windows?

al45tair replied:
I'll talk to you on Slack about this part, I think, since I think there's some discussion to be had about exactly what we want here.

grynspan replied:
Mumble mumble `std::filesystem::path`?

al45tair replied:
No. Because we don't have the minimum system version requirement on Darwin.

al45tair committed:
Improve path lookup.

On Windows, swift_getRootPath() should return the root path (i.e. outside
the "bin" directory).

On Linux, the path for the runtime looks like
<swift-root>/lib/swift/linux/libswiftCore.so
rather than <swift-root>/lib/swift/libswiftCore.so as you might expect.

On macOS, the path for the runtime *can* have "macosx" in it if we're using the
library from the build path.  That's not how it gets installed when it's part of
the operating system, but it's how we build it.

Also make swift_getAuxiliaryExecutablePath() actually check that the executable
exists.  This does mean it can return nullptr now, but it also means that it
can search through the various locations it supports to try to find the file
you're after, which is more robust.

rdar://103071801

al45tair commented:
@swift-ci Please smoke test

al45tair commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,369 @@
+{
+  // SWIFT_ROOT overrides the path returned by this function
+  const char *swiftRoot = swift::runtime::environment::SWIFT_ROOT();
+  if (swiftRoot && *swiftRoot) {
```
That probably is better, agreed.

al45tair committed:
Style tweak and an extra comment.

Change _swift_initRootPath() to do an early return, which reduces indentation.

Add a comment to reinforce that we can't use GetFinalPathNameByHandle() in
_swift_initRuntimePath().

rdar://103071801

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
Support even more paths.

It turns out that the path to the runtime might also include the
architecture as well as the platform.  So we need to search

  <root>/lib/swift/<platform>
  <root>/lib/swift/<platform>/<arch>
  <root>/lib/swift
  <root>/lib/swift/<arch>
  <root>/bin/
  <root>/bin/<arch>
  <root>

in that order.

Hopefully this now covers all the possibilities.

rdar://103071801

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
This is going to be replaced by https://github.com/al45tair/swift/pull/8 (part of the PR stack for the initial drop of the backtracing work).
